nx/lib/nx.ex

```diff
@@ -1228,6 +1217,199 @@ defmodule Nx do
     raise ArgumentError, "diag/2 expects tensor of rank 1 or 2, got: #{inspect(tensor)}"
   end
 
+  @doc """
+  Extracts the diagonal of a 2D tensor.
+
+  Converse of `make_diagonal/2`.
+
+  ## Examples
+
+  Given a 2D tensor without offset:
+
+      iex> Nx.take_diagonal(Nx.tensor([
+      ...> [0, 1, 2],
+      ...> [3, 4, 5],
+      ...> [6, 7, 8]
+      ...> ]))
+      #Nx.Tensor<
+        s64[3]
+        [0, 4, 8]
+      >
+
+  And if given a 2D tensor along with an offset:
+
+      iex> Nx.take_diagonal(Nx.iota({3, 3}), offset: 1)
+      #Nx.Tensor<
+        s64[2]
+        [1, 5]
+      >
+
+      iex> Nx.take_diagonal(Nx.iota({3, 3}), offset: -1)
+      #Nx.Tensor<
+        s64[2]
+        [3, 7]
+      >
+
+  ## Options
+
+    * `:offset` - offset used for extracting the diagonal.
+      Use offset > 0 for diagonals above the main diagonal,
+      and offset < 0 for diagonals below the main diagonal.
+
+  ## Error cases
+
+      iex> Nx.take_diagonal(Nx.tensor([0, 1, 2]))
+      ** (ArgumentError) take_diagonal/2 expects tensor of rank 2, got:
+      #Nx.Tensor<
+        s64[3]
+        [0, 1, 2]
+      >
+
+      iex> Nx.take_diagonal(Nx.iota({3, 3}), offset: 3)
+      ** (ArgumentError) offset must be less than breadth when positive, got: 3
+
+      iex> Nx.take_diagonal(Nx.iota({3, 3}), offset: -4)
+      ** (ArgumentError) absolute value of offset must be less than length when negative, got: -4
+  """
+  @doc type: :creation
```

Comment:
Is `:creation` the correct tag here and below? I am thinking it is rather `:shape`, because we are mostly keeping the same data but under a different shape? What do you think @polvalente?

Reply:
For `make_diagonal` I'd use `:creation`, since we can think of the operation as imposing certain elements of an already existing zero-matrix to have specific values. For me `:shape` would make sense if there weren't any other (zero) elements appearing in the resulting tensor.

Reply:
> For make_diagonal I'd use :creation, since we can think of the operation as imposing certain elements of an already existing zero-matrix to have specific values. For me :shape would make sense if there weren't any other (zero) elements appearing in the resulting tensor.

@polvalente, What about for `take_diagonal`?

Reply:
I think we can use the same tag for both. Although `take_diagonal` is more of "destruction" than "creation", because we eliminate elements from the matrix, but we can also think in terms of "fill this vector with elements from this matrix"

Reply:
So far all of our creation operations do not receive a tensor as input, only shape/parameters. So to me this is more of a transformation on the shape, but we can always revisit it later.